migrations/main.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"badi-cal/badi-events-backend/orm"
+	"badi-cal/badi-events-backend/terminal"
+	"fmt"
+	"log"
+
+	"gorm.io/gorm"
+)
+
+type SchemaMigration struct {
+	Migration string `gorm:"primaryKey"`
+}
+
+type MigrationAction func(db *gorm.DB)
+type Migration struct {
+	version string
+	up      MigrationAction
+	down    MigrationAction
+}
+
+// var argLength int = -1
+var versions []Migration
+
+func (migration *Migration) Migrate() {
+	db := orm.Connection()
+	var schema_migration SchemaMigration
+	if result := db.Where("migration = ?", migration.version).Take(&schema_migration); result.Error != nil {
+		fmt.Printf("Running migration '%s'", migration.version)
+		migration.up(db)
+		schema_migration = SchemaMigration{Migration: migration.version}
+		result := db.Create(&schema_migration)
+		if result.Error != nil {
+			log.Fatalf("Unable to mark migration. %v\n", result.Error)
+		}
+	} else {
+		fmt.Printf("Skipping migration '%s'; already run.\n", migration.version)
+		fmt.Printf("Error %v\n", result.Error)
+	}
+}
+
+func (migration *Migration) Rollback() {
+	db := orm.Connection()
+	var schema_migration SchemaMigration
+	if result := db.Where("migration = ?", migration.version).Take(&schema_migration); result.Error != nil {
+		fmt.Printf("Unable to rollback migration '%s'; has not been run.\n", migration.version)
+	} else {
+		fmt.Printf("Rolling back migration '%s'\n", migration.version)
+		migration.down(db)
+		db.Where("migration = ?", migration.version).Delete(&schema_migration)
+	}
+}
+
+func registerMigration(migration Migration) {
+	versions = append(versions, migration)
+}
+
+func main() {
```

Comment:
I feel a little spoiled in the Java world where I use a library from Flyway DB that does all of this for me to migrate database changes by just adding SQL scripts and sometimes Java classes with specific names to specific directories and then those SQL scripts and Java code get run for me automatically whenever the Java application starts up.  I guess there is not something like that for Go.

Reply:
I haven't found anything like that, but its there for every other language so its a little baffling.